uamqp/error.py

```diff
@@ -8,10 +8,133 @@
 from collections import namedtuple
 import six
 
-from .constants import PORT, FIELD
+from .constants import SECURE_PORT, FIELD
 from .types import AMQPTypes, FieldDefinition
 
 
+class ErrorCodes(Enum):
+    InternalError = b"amqp:internal-error"
+    IllegalState = b"amqp:illegal-state"
+    DecodeError = b"amqp:decode-error"
+    NotFound = b"amqp:not-found"
+    NotImplemented = b"amqp:not-implemented"
+    NotAllowed = b"amqp:not-allowed"
+    InvalidField = b"amqp:invalid-field"
+    ResourceLocked = b"amqp:resource-locked"
+    ResourceDeleted = b"amqp:resource-deleted"
+    UnauthorizedAccess = b"amqp:unauthorized-access"
+    FrameSizeTooSmall = b"amqp:frame-size-too-small"
+    ResourceLimitExceeded = b"amqp:resource-limit-exceeded"
+    PreconditionFailed = b"amqp:precondition-failed"
+    ConnectionRedirect = b"amqp:connection:redirect"
+    ConnectionCloseForced = b"amqp:connection:forced"
+    ConnectionFramingError = b"amqp:connection:framing-error"
+    SessionWindowViolation = b"amqp:session:window-violation"
+    SessionErrantLink = b"amqp:session:errant-link"
+    SessionHandleInUse = b"amqp:session:handle-in-use"
+    SessionUnattachedHandle = b"amqp:session:unattached-handle"
+    LinkRedirect = b"amqp:link:redirect"
+    LinkStolen = b"amqp:link:stolen"
+    LinkDetachForced = b"amqp:link:detach-forced"
+    LinkTransferLimitExceeded = b"amqp:link:transfer-limit-exceeded"
+    LinkMessageSizeExceeded = b"amqp:link:message-size-exceeded"
+    ClientError = b"amqp:client-error"
+    UnknownError = b"amqp:unknown-error"
+    VendorError = b"amqp:vendor-error"
+
+
+class RetryMode(str, Enum):
+    Exponential = 'exponential'
+    Fixed = 'fixed'
+
+
+class ErrorPolicy:
```

Comment:
I think we should try and refactor out the error policy in favour of a retry policy. Where errors are either retryable or they're not.

Reply:
is it about having two separate policies like:
```python
class ErrorPolicy:
# in the error policy, we define for each error code we do what
    self.error_map = {
        "some_error_code": (retryable, backoff_time)
    }

class RetryPolicy:
# we perform the retry based on the ErrorPolicy
try:

except Exception as e:
    action = error_policy.get_action(e.condtion)
    # do action

Reply:
hmm this seems a bit complicated - the HTTP retry policy has things like lists of status codes that it acts on etc, which is effectively what we're doing here.

Reply:
are you referring to the [retry settings](https://github.com/Azure/azure-sdk-for-python/blob/main/sdk/core/azure-core/azure/core/pipeline/policies/_retry.py#L86) as the ErrorPolicy?

I was initially thinking that `ErrorPolicy` and `RetryPolicy` are the same thing, but in azure-core there's no `ErrorPolicy` just retry policy taking some settings.

We could discuss about it more in tomorrow's sync up -- whether it's more of a design thing/what it should be looking like

Reply:
Yup more or less - though in the case of Http retry policy - it's largely a white list of status codes that it's safe to retry on.